### DISCUSSION TITLE
Melhor forma de construir relatórios em um SaaS

### DISCUSSION THREAD

**Discussion by kavalcante** [frontendbr/forum#1591] [Discussão]
Fala pessoal, tudo certo?

O Problema: Aqui na empresa, possuímos um SaaS com uma pegada de marketplace, onde temos 2 atores principais: compradores e fornecedores. Estamos querendo construir relatórios mais elaborados para cada um deles e para nós também. O problema é que existem muitas formas de fazer isso e queria entender como vocês fazem isso. Já vi alguns casos de uso e vou compartilhar com vocês algumas hipóteses nossas:

1. Construir as queries na mão e fazer os gráficos e etc... no front-end também.
- Pró: Controle total sobre o que entregamos.
- Contra: Maior trabalho possível

2. Utilizar o Metabase
- Pró: Muitas ferramentas para criar dashboard
- Contra: Não fiz uma POC, então não conheço a ferramenta. Mas já vi um caso de uso, onde você consegue passar os parametros encriptados via iframe (preciso mostrar o mesmo widget porém com informações diferentes para cada cliente)

3. Utilizar o Kibana
- Pró: Já utilizamos Elasticsearch na empresa
- Contra: Não sei se é possível exportar widgets igual o Metabase e normalmente o Kibana é para uso interno e não externos.

4. Utilizar outra ferramenta de BI
- Pró: Ferramentas mais conhecidas
- Contra: Preço

Vocês conhecem ou recomendam alguma coisa para resolver esse tipo de problema? Sugestões são muito bem vindas.

**Comment by victorferreira** [comment]
Lá na empresa passamos por discussões parecidas de tempos em tempos e sempre chegamos a mesma conclusão: Metabase, Kibana e coisas parecidas são ferramentas de uso interno, enquanto pra uso externo devemos construir uma solução customizada.

Embora pareça tentador usar uma solução de prateleira para servir os clientes tem alguns pontos a serem considerados quando vamos adotar uma ferramenta que não temos o domínio completo. Às vezes dá mais trabalho customizar o que está pronto do que construir algo do zero. Alguns dos pontos que acho importantes a serem considerados.

- Como funciona a integração com autenticação? É possível reutilizar o mesmo login e senha (ou seja lá o que vocês usem para autenticar um usuário)?

- Uma vez autenticado, como funcionam as permissões de acesso aos dados? Todo mundo pode ver tudo? Dá pra compartilhar as permissões?

- Como fica a consistência entre as interfaces que já existem? Customizar a interface é importante? É possível customizar a interface dessas ferramentas? Mesmo se for possível, vai dar muito trabalho? O que é preciso aprender?

- Como é o processo de migração caso seja decidido mudar de ferramenta ou mesmo desenvolver algo do zero?

- Qual o impacto da ferramenta na infraestrutura existente? Qual o nível de manutenção exigido pela ferramenta?

Por outro lado, embora construir tudo na mão possa ser bastante trabalhoso, existem muitas ferramentas que podem ajudar bastante nesse desenvolvimento.

- Independentemente do framework que vocês utilizem o que não falta por aí é _libs_ de gráficos, mas em última instancia sempre tem o D3. 

- Construir uma nova interface pode ser uma boa oportunidade pra reutilizar muito do código e ideias já utilizadas para construir o próprio SaaS. Quem sabe não sai daí um _design system_?

Algumas questões mais gerais também devem ser levadas são tão importantes quanto (talvez até mais):

- Qual a urgência dessa ferramenta? _Time to market_ pode ser crucial.

- Qual o orçamento? Quanto tempo e dinheiro podem ser gastos?

No fim das contas a resposta é sempre depende, mas uma coisa que pesou bastante na hora de escolher por construir algo do zero foi a possibilidade de fazer aos poucos. Fazer, testar e refinar pode levar um resultado que nenhuma outra ferramenta é capaz de oferecer. Espero ter ajudado.
